Adds an alternative python workflow generation path

danielabdi-noaa opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR provides an alternative workflow generation path using python3.

## TESTS CONDUCTED: 

### Hera
All W2E tests have been run on Hera using the python workflow generator and the existing shell script workflow generator.
Although there were about 24 failures, both workfkow generation paths result in the same tests successfully completed.

### Orion
All W2E tests have been run on Hera and there are 21 failures, even better than Hera.

### Jet
On hold. Jet seems to have a lot of missing data, especially with latest develop, which could not even generate the workflow
because of missing data:
```
EXTRN_MDL_SOURCE_BASEDIR_ICS = "/mnt/lfs4/BMC/wrfruc/UFS_SRW_app/staged_extrn_mdl_files/FV3GFS/nemsio"
```

### Unit tests
Unittest on the new files can be done as follows
```
python3 -m unittest -b \
	check_ruc_lsm.py \
	create_diag_table_file.py \
	create_model_configure_file.py \
	link_fix.py \
	set_cycle_dates.py \
	set_extrn_mdl_params.py \
	set_FV3nml_sfc_climo_filenames.py \
	set_FV3nml_ens_stoch_seeds.py \
	set_gridparams_ESGgrid.py \
	set_gridparams_GFDLgrid.py \
	set_ozone_param.py \
	set_predef_grid_params.py \
	set_thompson_mp_fix_files.py \
        get_crontab_contents.py

..............
----------------------------------------------------------------------
Ran 14 tests in 1.448s

OK
```
## DEPENDENCIES:
None

## DOCUMENTATION:
None

## ISSUE (optional): 
None

## CONTRIBUTORS (optional): 
@christinaholtNOAA for suggestion

christinaholtNOAA commented on ush/set_FV3nml_sfc_climo_filenames.py:
```diff
@@ -0,0 +1,139 @@
+        (_,nml_var_name,_) = run_command( f''' printf "%s\n" "{mapping}" | 
+                  {SED} -n -r -e "s/{regex_search}/\\1/p" ''')
+        (_,sfc_climo_field_name,_) = run_command(f''' printf "%s\n" "{mapping}" | 
+                          {SED} -n -r -e "s/{regex_search}/\\2/p" ''')
```
These lines might be able to leverage native python calls so we can more easily capture errors than with bash calls outside the python environment.

Also, I don't see `i` used here as a variable. Maybe `enumerate` isn't needed?

christinaholtNOAA commented on ush/set_FV3nml_sfc_climo_filenames.py:
```diff
@@ -0,0 +1,139 @@
+
+        fp = f'{FIXLAM}/{CRES}.{sfc_climo_field_name}.{suffix}'
+        if RUN_ENVIR != "nco":
+            (_,fp,_) = run_command( f'realpath --canonicalize-missing --relative-to="{dummy_run_dir}" "{fp}"' )
```
Can this path be built with python-native calls?

christinaholtNOAA commented on ush/set_cycle_dates.py:
```diff
@@ -0,0 +1,57 @@
+        if incr_cycl_freq % 24 != 0:
+            print_err_msg_exit(f'''
+                INCR_CYCL_FREQ is not divided by 24:
+                  INCR_CYCL_FREQ = \"{incr_cycl_freq}\"''')
```
@gsketefian Why is cycling frequency bound to be daily?

gsketefian replied:
I will defer to @chan-hoo on this one since he added this bit.

christinaholtNOAA commented on ush/set_ozone_param.py:
```diff
@@ -0,0 +1,233 @@
+      if symlink == ozone_symlink:
+        regex_search="^[ ]*([^| ]+[ ]*)[|][ ]*([^| ]*)[ ]*$"
+        (_,mapping_ozone,_)=run_command(f'''printf "%s\n" "{mapping}" | \
+                         {SED} -n -r -e "s/{regex_search}/\\1/p"''')
```
Repeat from a previous comment...can these leverage python-native commands instead of subprocesses?

christinaholtNOAA commented on ush/set_predef_grid_params.py:
```diff
@@ -0,0 +1,1214 @@
+#!/usr/bin/env python3
+
```
I'd love to see all this move to a YAML file, but may be outside the scope of this PR.

christinaholtNOAA reviewed (changes requested):
@danielabdi-noaa Thank you for this huge effort! I have gotten about halfway through a review, and a few of the comments I've made so far are applicable in many places within the code. In general I think it is useful to use Python-native commands where we can so that we can more easily leverage exception handling. We may also find that it will be faster than spinning up subprocesses to do things Python will handle for us (mostly untested conjecture). 

I have not finished my review just yet. I only made it through the `ush/set_thompson_mp_fix_files.py` and need to step away from it for a little while (maybe until Monday).

danielabdi-noaa commented on ush/create_diag_table_file.py:
```diff
@@ -0,0 +1,80 @@
+            cres: {CRES}''')
+
+    #call fill jinja
+    (err,_,_) = run_command( f'''{USHDIR}/fill_jinja_template.py -q -u "{settings}" -t {DIAG_TABLE_TMPL_FP} -o {diag_table_fp}''')
```
I have replaced all instances of such use of `fill_jinja_template` and `set_namelist` with a direct python function call. The arguments are passed a list of strings since that is the easiest thing to do. I could not pass it as one long string and break it down because of `settings` etc that are multiline strings... Let me know if you think of a better approach.

christinaholtNOAA replied:
Thanks for doing that. I was testing it out in a Python interactive session and finding that I wasn't able to just pass the list to the imported module. I needed these steps: 

```
import sys
### Probably want to save the original sys.argv here, and reset it after this step
sys.argv = ["-q", "-u", settings_str, "-t", DIAG_TABLE_TMPL_FP, "-o", diag_table_fp]
cla = fill_jinja_template.parse_args()
fill_jinja_template.main(cla)
```

We could certainly open PRs to the standalone tools to make them easier to deal with as imported scripts, if that's helpful here. Let me know what you think.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -0,0 +1,2157 @@
+          FIXgsm = \"{FIXgsm:-\"\"}
+          FIXaer = \"{FIXaer:-\"\"}
+          FIXlut = \"{FIXlut:-\"\"}
+          TOPO_DIR = \"{TOPO_DIR:-\"\"}
```
Is this syntax for f-strings, or leftover from bash?

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -0,0 +1,2157 @@
+
+    # export env vars
+    EXPORTS = ["NH0","NH3","NH4"]
+    export_vars(env_vars = EXPORTS)
```
Sorry that this is so late in the game, but it only just occurred to me. What if `export_vars()` accepted a dictionary of `env_vars`...would that free up the requirements for Python globals? Here's a quick version of what I was thinking the call might look like:


```
NH0=0
NH3=3
NH4=4
exports = dict(NHO=NHO, NH3=NH3, NH4=NH4)
export_vars(env_vars = exports)

christinaholtNOAA reviewed (changes requested):
@danielabdi-noaa Thanks again for all this work. A few more similar comments below. I think I made it through all of it this time...

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -0,0 +1,2157 @@
+        print_info_msg(msg)
+    
+      if RUN_TASK_VX_GRIDSTAT == True or \
+         RUN_TASK_VX_GRIDSTAT == False:
```
This logic seems unnecessary.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -0,0 +1,2157 @@
+    
+      if RUN_TASK_MAKE_SFC_CLIMO == True or \
+         ( RUN_TASK_MAKE_SFC_CLIMO == False and \
+           SFC_CLIMO_DIR != nco_fix_dir ):
```
Since these variables are boolean values, this logic statement becomes fairly simple in python:

```
if RUN_TASK_MAKE_SFC_CLIMO or \
  ( not RUN_TASK_MAKE_SFC_CLIMO and SFC_CLIMO_DIR != nco_fix_dir):
```

The comparison to a bool isn't necessary.

christinaholtNOAA reviewed (approved):
@danielabdi-noaa This is looking really nice. I have just a couple of super minor comments below. 

I think things are ready.

gsketefian commented:
@danielabdi-noaa This generally looks good to me.  Since it's hard to look through every line of this new code, it would be nice to at least try it out.  Do you have any instructions on how to run the python version of the workflow generation?  Say I want to run a single experiment with a given experiment configuration file.

danielabdi-noaa commented:
> 

@gsketefian To use the python workflow generation path, the only thing different is you execute `generate_FV3LAM_wflow.py` instead of the corresponding shell script file `generate_FV3LAM_wflow.sh`. The python workflow path is setup to access a shell script config file `config.sh` by default, even though it can use YAML format too, so there is no other changes you need to make.
It may be better to run both workflow patchs for a specific test case to compare the files generated (xml file, nml file etc). I've run the full regression test on Hera after modifying `run_W2E2_tests.sh` to use `generate_FV3LAM_wflow.py` and got same failures and passes as the shell script path. It would be good to have the test run on other systems.

gsketefian commented:
@danielabdi-noaa Ok thanks.  I will try it out on Hera and at least one more platform.

gsketefian commented:
@danielabdi-noaa I've been running tests on both Hera and Orion.  I decided that it's best (and enough) to run the `wflow_features` tests (i.e. no need to run the tests in `grids_extrn_mdls_suites_community` and `grids_extrn_mdls_suites_nco`).  There are a total of 35 of those.  I've run 10 tests on Hera (including the bash counterparts) with the same results as with bash (2 failures), so I'm not worried about the failures so far.  I'm planning on running the rest today.

I've also run 6 tests on Orion with some failures, although I haven't run the bash counterparts there yet.  I had to make some adjustments to get things to run on Orion, but nothing directly related to this PR.  I'm working through those failures now and will do more tests today.

One thing I noticed is that for a pre-existing file or directory that needs to be renamed, the python version adds the date and time to the end of the name instead of "oldNN".  That is good since it identifies when it was renamed.  But can we also add the string "old" somewhere in the rename?  E.g. instead of renaming `specify_DOT_OR_USCORE` to say `specify_DOT_OR_USCORE_20220407_100006`, rename it to `specify_DOT_OR_USCORE_old_20220407_100006`.  That's just to make it clear that it's an older copy.  

I'll keep you posted with more results soon.  Thanks.

danielabdi-noaa commented:
@gsketefian Thank you for putting the effort to test it, I know it is an unusually big PR!
I will make the changes you suggested for renaming old folders.
One thing I should mention is that the python code could be a couple of commits behind so that could make some difference.
It was upto date last week but haven't checked since then.

I just got access to the wrfruc project jet and niagra so I will do the rull regression tests there too.
I've run a couple of tests on jet and looks to run there just fine.

Thanks again!

gsketefian commented:
@danielabdi-noaa You're welcome.  And sorry it takes a while for me to get to these big PRs.

I did notice that the python scripts are not completely up-to-date, so I'm testing with slightly older hashes.  In particular, I am not using the latest hashes that went in PR #[233](https://github.com/ufs-community/ufs-srweather-app/pull/233).  If you want to update to the latest, that would be great (I guess you'll have to do that anyway before merging).  Let me know if you do.  There are some tests (specifically with the `GFS_2017_gfdlmp[_regional]` suite that are failing that would be fixed if you merged in the latest from `develop`.

Also, on Orion, I need to update the data that's available to run several of the tests since it's not up-to-date with the latest data.

gsketefian commented:
@danielabdi-noaa One thing I'm finding on Orion is that after a test finishes (either successfully or not), the cron job for it is not removed from the user's cron table.  The simplest case to demo this is with the `deactivate_tasks` test (which completes successfully).  I ran it with the bash scripts, and the cron job was removed.  Then I tried with python, and it wasn't.  

Do you have an account on Orion?  I imagine it would be hard to debug if you don't.  The interesting thing is that this doesn't happen on Hera, which is very similar to Orion.

Note that on Orion, you have to be on login node 1 to be able to use `cron`.  I can provide more details about this if you like.

danielabdi-noaa commented:
> @danielabdi-noaa One thing I'm finding on Orion is that after a test finishes (either successfully or not), the cron job for it is not removed from the user's cron table. The simplest case to demo this is with the `deactivate_tasks` test (which completes successfully). I ran it with the bash scripts, and the cron job was removed. Then I tried with python, and it wasn't.
> 
> Do you have an account on Orion? I imagine it would be hard to debug if you don't. The interesting thing is that this doesn't happen on Hera, which is very similar to Orion.
> 
> Note that on Orion, you have to be on login node 1 to be able to use `cron`. I can provide more details about this if you like.

I haven't used my Orion account in a long time, but I will try to reactivate it and see if I can reproduce the problem.
Thanks

gsketefian commented:
@danielabdi-noaa I imagine this PR will affect the documentation that EPIC is working on one way or another.  I am tagging @gspetro-NOAA , who's doing the documentation for the release.

danielabdi-noaa commented:
@gsketefian  I have got access to Orion now. Please let me know how I can get login node 1 because all I got was login node 2 which doesn't allow a cron job.

I have already fixed one bug where the temp files `/tmp/t1` and `/tmp/t2` needed for extracting variables set in shell script are shared by different users. You already created those so I was not even able to generate the workflow before the fix ...

gsketefian commented:
@danielabdi-noaa Ahh, I see.  Ok, one less bug then!  I used to be able to login directly into node 1 from my laptop, but that's no longer possible.  I first login to wherever it puts me (usually it seems to be node 2), then I issue this:
```
ssh -Y gketefia@orion-login-1.hpc.msstate.edu
```

danielabdi-noaa commented:
@gsketefian Thanks I am able to reproduce the problem now. It looks like deactive_tasks behave similarly on Hera too i.e. a cronjob line is left hanging after job is done. It should be easy to debug now ...

danielabdi-noaa commented:
@gsketefian I have fixed the crontab issue now, which was due to double quotes being lost in `called_from_cron="TRUE"` when the crontab line is written. So at the end of workflow run  `grep -v` was not able to match the line with CRONTAB_LINE so it was not able to remove it.

Question, there is this line in the workflow generating script which implies that `called_from_cron` could be set during workflow generation.
https://github.com/ufs-community/regional_workflow/blob/3f83cde36876741ec5cf249e70c3c5ee5fd699ec/ush/generate_FV3LAM_wflow.sh#L505
I have it to always take the default value in the python workflow path because I couldn't figure out where it could be set.

gsketefian commented:
@danielabdi-noaa FYI the `called_from_cron` is needed as an argument to the workflow (re)launch script `launch_FV3LAM_wflow.sh` because of the behavior of `crontab` on Cheyenne.  That's described in more detail in PR #675.  From that PR:

> Adds an optional argument to launch_FV3LAM_wflow.sh that specifies whether it is being called from a cron job. This is necessary because on Cheyenne, the version of crontab to use depends on whether or not the parent script is being called from a cron job.

Anyway, the only place that `called_from_cron` is set to "TRUE" is when `launch_FV3LAM_wflow.sh` is called from the command line.  You can see that from here:
https://github.com/ufs-community/regional_workflow/blob/3f83cde36876741ec5cf249e70c3c5ee5fd699ec/ush/setup.sh#L1441
So that argument is only set internally when a cron job line is being created that will then be written to the user's cron table.  Otherwise, the default value (FALSE) should be used.  So you should be ok, I think.

Can I now go ahead and update my version with your latest?  Thanks.

danielabdi-noaa commented:
@gsketefian Thanks for the explanation.

> Can I now go ahead and update my version with your latest? Thanks.

Yes, you can try it out now. I have also updated the PR to the latest develop but it is not well tested.
It did run the `deactivate_tasks` properly on Orion though.

Note that I rebased the PR to develop and force pushed so you may need to make a clean clone of the PR.

danielabdi-noaa commented:
@gsketefian I think Orion is working great now, maybe even better than Hera, so I am putting it in the "tested section".
I have run the whole regression test and got fewer failures than Hera's (only 20).
Moreover, there were no crontab lines hanging around once the test suite finished running.
My test directory on Orion is at:
```
/work/noaa/gsd-hpcs/Daniel.Abdi/benchmark-run
```
Jet, on the other hand, seems to have a lot of missing data especially with latest develop so I am going to put that test on hold ...

willmayfield commented:
@danielabdi-noaa Thanks for your work this is great!

I tried generating and running the GST case by hand on Cheyenne and ran into some minor issues: 1. while bringing in environment variables, for some reason the (NPL) from the Cheyenne workflow environment prompt ends up causing an error. 2. When EXPT_BASEDIR is an empty string in the configs (as it is by default) indexing EXPT_BASEDIR[0] gives an error in setup.py. 3. MACHINE="cheyenne" fails when it gets compared to the upper-case machine names list in valid_param_vals.yaml.

I also tried it on Orion and still got the same problems 2 and 3 there. If I avoid triggering these checks, both generated and ran successfully!

danielabdi-noaa commented:
@willmayfield Thank you for testing and finding out issues -- I will try and fix them later!

danielabdi-noaa commented:
> @danielabdi-noaa Thanks for your work this is great!
> 
> I tried generating and running the GST case by hand on Cheyenne and ran into some minor issues: 1. while bringing in environment variables, for some reason the (NPL) from the Cheyenne workflow environment prompt ends up causing an error. 2. When EXPT_BASEDIR is an empty string in the configs (as it is by default) indexing EXPT_BASEDIR[0] gives an error in setup.py. 3. MACHINE="cheyenne" fails when it gets compared to the upper-case machine names list in valid_param_vals.yaml.
> 
> I also tried it on Orion and still got the same problems 2 and 3 there. If I avoid triggering these checks, it generated and ran successfully on both platforms!

@willmayfield  I believe that I 've fixed the last two issues but can not test the Cheyenne issue because I do not have access to the machine.
I was wondering if you can give me more information to debug the problem. There is a brittle section of code in export/import vars that could be the cause of the problem:

https://github.com/danielabdi-noaa/regional_workflow/blob/python_workflow/ush/python_utils/environment.py#L239-L245

I skip python functions, lowercase environment variables that are unlikely to be variables defined by SRW app, which is what I want to import/export. It could be that something on cheyenne made it through and is causing problems?

JeffBeck-NOAA commented:
@willmayfield, do you have time to test @danielabdi-noaa's changes to this PR on Cheyenne?  @gsketefian, does this need to be retested on Hera?

danielabdi-noaa commented:
> @willmayfield, do you have time to test @danielabdi-noaa's changes to this PR on Cheyenne? @gsketefian, does this need to be retested on Hera?

@JeffBeck-NOAA I believe Will has tested and confirmed my latest changes fix the problem on Cheyenne. Our communications shifted to email a couple of days ago.

JeffBeck-NOAA commented:
@danielabdi-noaa, thanks!  I will go ahead and approve.  @gsketefian, did you want to test one more time on Hera?

venitahagerty commented:
Machine: jet
Compiler: intel
Job: WE
Repo location: /lfs1/BMC/nrtrr/rrfs_ci/autoci/pr/876620093/20220425230518/ufs-srweather-app
Build was Successful
Rocoto jobs started
Experiment failed: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
2022-04-25 23:42:15 +0000 :: fe2 :: Task get_extrn_ics, jobid=2642517, in state DEAD (FAILED), ran for 15.0 seconds, exit status=256, try=1 (of 1)
Long term tracking will be done on 9 experiments
If test failed, please make changes and add the following label back:
ci-jet-intel-WE
